v2/sourcedb-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/PipelineController.java

```diff
@@ -248,6 +289,11 @@ static PipelineResult executeShardedMigration(
     return pipeline.run();
   }
 
+  // Calculate the total number of logical shards in the list of physical shards.
+  private static long findNumLogicalshards(List<Shard> shards) {
+    return shards.stream().mapToLong(shard -> shard.getDbNameToLogicalShardIdMap().size()).sum();
```

Comment:
is this null safe ? is it being checked during parsing of the shards ?

Reply:
Yep this reaching this stage means the list is not null and each elements has atleast one logical shard.